File: ocaml/lib/program_generation.ml
type ident = string

type envcomp =
  | Var of ident
  | Lam of ident * envcomp
  | App of envcomp * envcomp
  | LamD of ident * envcomp
  | AppD of envcomp * envcomp

type exp =
  | Var' of ident
  | Lam' of ident * exp
  | App' of exp * exp
  | LamD' of ident * exp
  | AppD' of exp * exp

type value =
  | Fun of (value -> value)
  | Code of exp

let rec spec e r =
  match e with
  | Var' x -> lookup r x
  | Lam' (x, e) -> Fun (fun y -> spec e (upd r x y))
  | App' (f, a) ->
     let Fun ff = spec f r in
     ff (spec a r)
  | LamD' (x, e) ->
     let xx = gensym () in
     let Code body = spec e (upd r x (Code (Var' xx))) in
     Code (Lam' (xx, body))
  | AppD' (f, a) ->
     let Code ff = spec f r in
     let Code aa = spec a r in
     Code (App' (ff, aa))

File: rust/src/program_generation.rs
use std::collections::HashMap;

type Ident = String;

enum Envcomp {
    Var(Ident),
    Lam(Ident, Box<Envcomp>),
    App(Box<Envcomp>, Box<Envcomp>),
    LamD(Ident, Box<Envcomp>),
    AppD(Box<Envcomp>, Box<Envcomp>),
}

enum Exp {
    Var'(Ident),
    Lam'(Ident, Box<Exp>),
    App'(Box<Exp>, Box<Exp>),
    LamD'(Ident, Box<Exp>),
    AppD'(Box<Exp>, Box<Exp>),
}

enum Value {
    Fun(Box<dyn Fn(Value) -> Value>),
    Code(Box<Exp>),
}

fn spec(e: &Exp, r: &mut HashMap<Ident, Value>) -> Value {
    match e {
        Exp::Var'(x) => r.get(x).cloned().unwrap_or(Value::Code(Box::new(Exp::Var'(x.clone())))),
        Exp::Lam'(x, e) => Value::Fun(Box::new(move |y| {
            r.insert(x.clone(), y);
            spec(e, r)
        })),
        Exp::App'(f, a) => match (spec(f, r), spec(a, r)) {
            (Value::Fun(ff), a) => ff(a),
            _ => panic!("Type error"),
        },
        Exp::LamD'(x, e) => {
            let xx = gensym();
            r.insert(x.clone(), Value::Code(Box::new(Exp::Var'(xx.clone()))));
            let body = spec(e, r);
            Value::Code(Box::new(Exp::Lam'(xx, Box::new(match body {
                Value::Code(e) => *e,
                _ => panic!("Type error"),
            }))))
        }
        Exp::AppD'(f, a) => match (spec(f, r), spec(a, r)) {
            (Value::Code(ff), Value::Code(aa)) => Value::Code(Box::new(Exp::App'(ff, aa))),
            _ => panic!("Type error"),
        },
    }
}